net/mock index out of range panic

I see an index out of range panic within `p2p/net/mock:ConnectAllButSelf()` very occasionally when running the code below a bunch of times.  This code creates a mocknet of size 2, links all hosts together and then has each host connect to all other hosts but themselves.

```golang
package mocknettest                                                                                                                                           
                                                                                                                                                              
import (                                                                                                                                                      
        "context"                                                                                                                                             
        "testing"                                                                                                                                             
        "time"                                                                                                                                                
                                                                                                                                                              
        mocknet "gx/ipfs/QmXnpYYg2onGLXVxM4Q5PEFcx29k8zeJQkPeLAk9h9naxg/go-libp2p/p2p/net/mock"                                                               
                                                                                                                                                              
        "github.com/stretchr/testify/require"                                                                                                                 
)                                                                                                                                                             
                                                                                                                                                              
func TestMockNet(t *testing.T) {                                                                                                                              
        t.Parallel()                                                                                                                                          
        ctx, cancel := context.WithCancel(context.Background())                                                                                               
        defer cancel()                                                                                                                                        
        require := require.New(t)                                                                                                                             
        mn, err := mocknet.WithNPeers(ctx, 2)                                                                                                                 
        require.NoError(err)                                                                                                                                  
                                                                                                                                                              
        require.NoError(mn.LinkAll())                                                                                                                         
        require.NoError(mn.ConnectAllButSelf())                                                                                                               
                                                                                                                                                              
        time.Sleep(time.Millisecond * 50)                                                                                                                     
}                                                                                                                                                             
```

The stack trace:
```
=== RUN   TestMockNet
=== PAUSE TestMockNet
=== CONT  TestMockNet
--- FAIL: TestMockNet (0.00s)
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 793540 [running]:
testing.tRunner.func1(0xc00040ef00)
	/usr/local/go/src/testing/testing.go:792 +0x387
panic(0x14aed60, 0x1abe4a0)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
gx/ipfs/QmXnpYYg2onGLXVxM4Q5PEFcx29k8zeJQkPeLAk9h9naxg/go-libp2p/p2p/net/mock.newConn(0xc0005a6600, 0xc0005a6680, 0xc00027e7d0, 0x2, 0x8)
	/Users/zenground0/Ground/go/src/gx/ipfs/QmXnpYYg2onGLXVxM4Q5PEFcx29k8zeJQkPeLAk9h9naxg/go-libp2p/p2p/net/mock/mock_conn.go:44 +0x481
gx/ipfs/QmXnpYYg2onGLXVxM4Q5PEFcx29k8zeJQkPeLAk9h9naxg/go-libp2p/p2p/net/mock.(*link).newConnPair(0xc00027e7d0, 0xc0005a6600, 0x0, 0x0)
	/Users/zenground0/Ground/go/src/gx/ipfs/QmXnpYYg2onGLXVxM4Q5PEFcx29k8zeJQkPeLAk9h9naxg/go-libp2p/p2p/net/mock/mock_link.go:36 +0xb9
gx/ipfs/QmXnpYYg2onGLXVxM4Q5PEFcx29k8zeJQkPeLAk9h9naxg/go-libp2p/p2p/net/mock.(*peernet).openConn(0xc0005a6600, 0xc0004d73e2, 0x7, 0xc00027e7d0, 0x2)
	/Users/zenground0/Ground/go/src/gx/ipfs/QmXnpYYg2onGLXVxM4Q5PEFcx29k8zeJQkPeLAk9h9naxg/go-libp2p/p2p/net/mock/mock_peernet.go:155 +0x3c
gx/ipfs/QmXnpYYg2onGLXVxM4Q5PEFcx29k8zeJQkPeLAk9h9naxg/go-libp2p/p2p/net/mock.(*peernet).connect(0xc0005a6600, 0xc0004d73e2, 0x7, 0x2, 0x2, 0x0)
	/Users/zenground0/Ground/go/src/gx/ipfs/QmXnpYYg2onGLXVxM4Q5PEFcx29k8zeJQkPeLAk9h9naxg/go-libp2p/p2p/net/mock/mock_peernet.go:150 +0x3a8
gx/ipfs/QmXnpYYg2onGLXVxM4Q5PEFcx29k8zeJQkPeLAk9h9naxg/go-libp2p/p2p/net/mock.(*peernet).DialPeer(0xc0005a6600, 0x16d4ae0, 0xc00023d240, 0xc0004d73e2, 0x7, 0x0, 0x2000100, 0x0, 0xffffffffffffffff)
	/Users/zenground0/Ground/go/src/gx/ipfs/QmXnpYYg2onGLXVxM4Q5PEFcx29k8zeJQkPeLAk9h9naxg/go-libp2p/p2p/net/mock/mock_peernet.go:117 +0x3f
gx/ipfs/QmXnpYYg2onGLXVxM4Q5PEFcx29k8zeJQkPeLAk9h9naxg/go-libp2p/p2p/net/mock.(*mocknet).ConnectNets(0xc00027e3c0, 0x16d81c0, 0xc0005a6600, 0x16d81c0, 0xc0005a6680, 0x0, 0x0, 0x0, 0x0)
	/Users/zenground0/Ground/go/src/gx/ipfs/QmXnpYYg2onGLXVxM4Q5PEFcx29k8zeJQkPeLAk9h9naxg/go-libp2p/p2p/net/mock/mock_net.go:329 +0x74
gx/ipfs/QmXnpYYg2onGLXVxM4Q5PEFcx29k8zeJQkPeLAk9h9naxg/go-libp2p/p2p/net/mock.(*mocknet).ConnectAllButSelf(0xc00027e3c0, 0x0, 0x0)
	/Users/zenground0/Ground/go/src/gx/ipfs/QmXnpYYg2onGLXVxM4Q5PEFcx29k8zeJQkPeLAk9h9naxg/go-libp2p/p2p/net/mock/mock_net.go:316 +0x1a3
practice/libp2p-bug.TestMockNet(0xc00040ef00)
	/Users/zenground0/Ground/go/src/practice/libp2p-bug/main_test.go:22 +0x14a
testing.tRunner(0xc00040ef00, 0x1674d38)
	/usr/local/go/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:878 +0x353
exit status 2
FAIL	practice/libp2p-bug	1335.435s
```

--- a/p2p/net/mock/mock_net.go
+++ b/p2p/net/mock/mock_net.go
@@ -3,6 +3,7 @@ package mocknet
 import (
 	"context"
 	"fmt"
+	"net"
 	"sort"
 	"sync"
 
@@ -17,9 +18,12 @@ import (
 	peer "github.com/libp2p/go-libp2p-peer"
 	pstore "github.com/libp2p/go-libp2p-peerstore"
 	pstoremem "github.com/libp2p/go-libp2p-peerstore/pstoremem"
-	testutil "github.com/libp2p/go-testutil"
 	ma "github.com/multiformats/go-multiaddr"
 )
+
+// IP6 range that gets blackholed (in case our traffic ever makes it out onto
+// the internet).
+var blackholeIP6 = net.ParseIP("100::")
 
 // mocknet implements mocknet.Mocknet
 type mocknet struct {
@@ -54,8 +58,20 @@ func (mn *mocknet) GenPeer() (host.Host, error) {
 	if err != nil {
 		return nil, err
 	}
-
-	a := testutil.RandLocalTCPAddress()
+	id, err := peer.IDFromPrivateKey(sk)
+	if err != nil {
+		return nil, err
+	}
+	suffix := id
+	if len(id) > 8 {
+		suffix = id[len(id)-8:]
+	}
+	ip := append(net.IP{}, blackholeIP6...)
+	copy(ip[net.IPv6len-len(suffix):], suffix)
+	a, err := ma.NewMultiaddr(fmt.Sprintf("/ip6/%s/tcp/4242", ip))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create test multiaddr: %s", err)
+	}
 
 	h, err := mn.AddPeer(sk, a)
 	if err != nil {

--- a/p2p/net/mock/mock_test.go
+++ b/p2p/net/mock/mock_test.go
@@ -582,6 +582,14 @@ func TestLimitedStreams(t *testing.T) {
 		t.Fatal("Expected 2ish seconds but got ", time.Since(before))
 	}
 }
+func TestFuzzManyPeers(t *testing.T) {
+	for i := 0; i < 50000; i++ {
+		_, err := FullMeshConnected(context.Background(), 2)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+}
 
 func TestStreamsWithLatency(t *testing.T) {
 	latency := time.Millisecond * 500

